aten/src/ATen/native/cuda/SegmentReduce.cu

```diff
@@ -0,0 +1,124 @@
+
+#include <ATen/native/SegmentReduce.h>
+
+#include <ATen/ATen.h>
+#include <ATen/NumericUtils.h>
+#include <ATen/cuda/CUDAContext.h>
+#include <ATen/cuda/detail/KernelUtils.h>
+#include <cub/cub.cuh>
+#include <iostream>
+
+namespace at {
+namespace native {
+
+struct CustomMax {
+  /// Boolean max operator, returns <tt>(a > b) ? a : b</tt>
+  template <typename OutputT>
+  __host__ __device__ __forceinline__ OutputT
+  operator()(const OutputT& a, const OutputT& b) {
+    return at::_isnan(b) ? b : std::max<OutputT>(a, b);
+  }
+};
+
+Tensor _get_complete_sum(const Tensor& lengths) {
+  int64_t segment_count = lengths.numel();
+  auto offsets = at::empty({segment_count + 1}, lengths.options());
+  offsets[0].zero_();
+  auto* lengths_data_ptr = lengths.data_ptr<int64_t>();
+  auto* offsets_data_ptr = offsets.data_ptr<int64_t>();
+  size_t temp_storage_bytes = 0;
+  cudaError_t error = cub::DeviceScan::InclusiveSum(
+      nullptr,
+      temp_storage_bytes,
+      lengths_data_ptr,
+      offsets_data_ptr + 1,
+      segment_count,
+      at::cuda::getCurrentCUDAStream());
+  TORCH_INTERNAL_ASSERT(error == cudaSuccess, "Failed to make cub call!");
+
+  auto offset_temp_storage = at::empty(
+      {static_cast<int64_t>(temp_storage_bytes)},
+      lengths.options().dtype(kByte));
+
+  error = cub::DeviceScan::InclusiveSum(
+      offset_temp_storage.data_ptr(),
+      temp_storage_bytes,
+      lengths_data_ptr,
+      offsets_data_ptr + 1,
+      segment_count,
+      at::cuda::getCurrentCUDAStream());
+  TORCH_INTERNAL_ASSERT(error == cudaSuccess, "Failed to make cub call!");
+
+  return offsets;
+}
+
+Tensor _segment_reduce_cuda_kernel(
+    const Tensor& data,
+    const c10::optional<Tensor>& lengths,
+    int64_t axis,
+    bool unsafe) {
+  const auto& lengths_value = lengths.value();
+  if (!unsafe) {
+    TORCH_CHECK(
+        (lengths_value >= 0).all().item<int64_t>(),
```

Comment:
length_value.min()>0 is faster (we've agreed that there shouldn't be 0-lengths segments, right?)